### PR TITLE
Delete lock file if contained pid is stale

### LINKED ISSUE #2: detect stale lock file
When a microcom process died but failed to remove the lock file (`/var/lock/LCK..$devname`), restarting microcom fails with

    lockfile for port exists

This could be improved by checking the content of the file and ignore its presence if the related pid doesn't exist.

### PR DESCRIPTION
**Pull request by a3f**
Fixes #2.

### CODE REVIEW

File: serial.c [hunk 1]
```diff
@@ -178,15 +180,18 @@ struct ios_ops * serial_init(char *device)
 	}
 
 	fd = open(lockfile, O_RDONLY);
-	if (fd >= 0 && !opt_force) {
+	if (fd >= 0) {
+		read(fd, &pid, sizeof(long));
```
[the remark this answers is not in the document]
Comment (ukleinek) [reply serial.c]: Another patch opportunity: The FHS demands this locking scheme:
http://www.pathname.com/fhs/pub/fhs-2.3.html#VARLOCKLOCKFILES

[the remark this answers is not in the document]
Comment (a3f) [reply serial.c]: Ok, will change it along with that note then:

https://github.com/pengutronix/microcom/blob/d82825c25a651c8202a7d836c6ca3612adf3e67f/serial.c#L200

**Comment by a3f** [comment]
I've changed microcom to use the HDB UUCP format for its lock files.

Deleting stale lock files though is racy and I think it can't be fixed without file locks, but as it was racy before anyway, it might still be good enough for inclusion :-)

**Comment by a3f** [comment]
Any comments on this?

File: serial.c [hunk 2]
```diff
@@ -177,26 +178,28 @@ struct ios_ops * serial_init(char *device)
...
-		main_usage(3, "lockfile for port exists", device);
-	}
+relock:
+	fd = open(lockfile, O_RDWR | O_CREAT | O_EXCL, 0444);
+	if (fd < 0) {
+		if (errno == EEXIST) {
+			if (opt_force) {
+				printf("lockfile for port exists, ignoring\n");
+				serial_unlock();
+				goto relock;
```
Comment (ukleinek) [inline serial.c]: if `serial_unlock()` fails to delete the lockfile, this is an endless loop.

File: serial.c [hunk 3]
```diff
@@ -182,12 +182,34 @@ struct ios_ops * serial_init(char *device)
...
 			}
 
+			fd = open(lockfile, O_RDONLY);
+			if (fd < 0)
+				main_usage(3, "lockfile for port can't be opened", device);
+
+			do {
+				ret = read(fd, &pidbuf[nbytes], sizeof(pidbuf) - nbytes - 1);
+				nbytes += ret;
+			} while (ret > 0 && nbytes < sizeof (pidbuf) - 1);
```
Comment (ukleinek) [inline serial.c]: Would it make sense here to do some additional checks on the format? E.g. check you read the file completely? If the lockfile happens to be created by an older microcom with pid 33333333 you check for pid 3333 instead. (I think, didn't test.)

File: serial.c [hunk 4]
```diff
@@ -182,12 +182,34 @@ struct ios_ops * serial_init(char *device)
...
+
+			if (ret >= 0) {
+				pidbuf[nbytes] = '\0';
+				ret = sscanf(pidbuf, "%10ld\n", &pid);
+
+				if (ret == 1 && kill(pid, 0) < 0 && errno == ESRCH) {
+					printf("lockfile contains stale pid, ignoring\n");
+					serial_unlock();
+					goto relock;
```
Comment (ukleinek) [inline serial.c]: same issue as with the first patch in this series (i.e. endless loop if serial_unlock fails to remove lockfile)

**Comment by a3f** [comment]
With move to flock, this is no longer applicable.

**Pull request closed without merging**